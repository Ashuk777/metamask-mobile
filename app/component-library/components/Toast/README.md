# Toast

Toast is a generic component to show temporary or dismissable messages to the user, it's a wrapper around [react-native-toast-message](https://github.com/calintamas/react-native-toast-message).

## Props

This component doesn't receive any props directly because it's intented to be triggered through a function call.

## Toast.showWarningToast

### `title`

The text to display in the toast title.

| <span style="color:gray;font-size:14px">TYPE</span>    | <span style="color:gray;font-size:14px">REQUIRED</span> |
| :----------------------------------------------------- | :------------------------------------------------------ |
| string | Yes                                                     | 

### `message`

The text to display in the toast body.

| <span style="color:gray;font-size:14px">TYPE</span>    | <span style="color:gray;font-size:14px">REQUIRED</span> |
| :----------------------------------------------------- | :------------------------------------------------------ |
| string | Yes                                                     | 

### `action`

The function to call when the action link is pressed.

| <span style="color:gray;font-size:14px">TYPE</span>    | <span style="color:gray;font-size:14px">REQUIRED</span> |
| :----------------------------------------------------- | :------------------------------------------------------ |
| string | Yes                                                     | 

### `actionText`

The text to display in the action link.

| <span style="color:gray;font-size:14px">TYPE</span>    | <span style="color:gray;font-size:14px">REQUIRED</span> |
| :----------------------------------------------------- | :------------------------------------------------------ |
| string | Yes                                                     | 

## Usage

```javascript
import Toast from '../../../component-library/components/Toast';

Toast.showWarningToast({
  title: 'Security Warning',
  message:
    'Your Android System Webview is out of date and may leave you open to security risks.',
  actionText: 'Update on Play Store',
  action: () => console.log('LOG'),
});
```
