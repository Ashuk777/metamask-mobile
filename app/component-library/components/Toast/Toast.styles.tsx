import { StyleSheet } from 'react-native';
import { Theme } from '../../../util/theme/models';

const HEIGHT = 60;
const WIDTH = 340;
const BORDER_RADIUS = 8;

/**
 * Style sheet function for Link component.
 *
 * @param params Style sheet params.
 * @param params.theme App theme from ThemeContext.
 * @param params.vars Inputs that the style sheet depends on.
 * @returns StyleSheet object.
 */
const styleSheet = (params: { theme: Theme }) => {
  const { theme } = params;
  return StyleSheet.create({
    baseToastContainer: {
      width: WIDTH,
      minHeight: HEIGHT,
      borderColor: theme.colors.warning.default,
      borderWidth: 1,
      borderRadius: BORDER_RADIUS,
      marginHorizontal: 10,
      backgroundColor: theme.colors.warning.inverse,
      shadowOffset: { width: 0, height: 0 },
      shadowOpacity: 0.1,
    },
    baseToastWrapper: {
      paddingHorizontal: 10,
      paddingVertical: 10,
      backgroundColor: theme.colors.warning.muted,
      flexDirection: 'row',
    },
    warningToastIconContainer: {
      marginTop: 3,
    },
    warningToastTextContainer: {
      flex: 1,
      marginLeft: 10,
    },
    warningToastTitle: {
      marginBottom: 5,
    },
    warningToastActionsContainer: {
      flexDirection: 'row',
      marginTop: 10,
    },
    warningToastdismissableAction: {
      marginRight: 15,
    },
  });
};

export default styleSheet;
