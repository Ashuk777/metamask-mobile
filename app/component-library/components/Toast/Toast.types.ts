export interface IToast extends React.FC {
  showWarningToast: (params: ShowWarningToastParams) => void;
}

export interface ShowWarningToastParams {
  /**
   * The text to display in the toast title.
   */
  title: string;
  /**
   * The text to display in the toast body.
   */
  message: string;
  /**
   * The function to call when the action link is pressed.
   */
  action: () => void;
  /**
   * The text to display in the action link.
   */
  actionText: string;
}

export interface WarningToastProps {
  /**
   * The function to call when the action link is pressed.
   */
  action: () => void;
  /**
   * The text to display in the action link.
   */
  actionText: string;
}
