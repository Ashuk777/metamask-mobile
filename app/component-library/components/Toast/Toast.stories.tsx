/* eslint-disable no-console */
import React, { useEffect } from 'react';
import { storiesOf } from '@storybook/react-native';
import Toast from './Toast';

const ToastWrapper = () => {
  useEffect(() => {
    Toast.showWarningToast({
      title: 'Security Warning',
      message:
        'Your Android System Webview is out of date and may leave you open to security risks.',
      actionText: 'Update on Play Store',
      action: () => console.log('LOG'),
    });
  }, []);

  return <Toast />;
};

storiesOf('Component Library / Toast', module)
  .addDecorator((getStory) => getStory())
  .add('Warning', () => <ToastWrapper />);
