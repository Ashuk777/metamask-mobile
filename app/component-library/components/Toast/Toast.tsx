import React from 'react';
import { View } from 'react-native';
import RNToastMessage from 'react-native-toast-message';
import Icons from 'react-native-vector-icons/MaterialCommunityIcons';
import { strings } from '../../../../locales/i18n';
import { useAppThemeFromContext } from '../../../util/theme';
import { useStyles } from '../../hooks';
import BaseText, { BaseTextVariant } from '../BaseText';
import Link from '../Link';
import styleSheet from './Toast.styles';
import { WarningToastProps, IToast } from './Toast.types';

const BaseToast = ({
  styles,
  children,
}: {
  styles: any;
  children: React.ReactElement;
}) => (
  <View style={styles.baseToastContainer}>
    <View style={styles.baseToastWrapper}>{children}</View>
  </View>
);

const toastConfig = (styles: any, colors: any) => ({
  warningToast: function warningToast({
    text1,
    text2,
    props,
  }: {
    text1?: string | undefined;
    text2?: string | undefined;
    props: WarningToastProps;
  }) {
    return (
      <BaseToast styles={styles}>
        <>
          <View style={styles.warningToastIconContainer}>
            <Icons name="bell-outline" size={21} color={colors.text.default} />
          </View>
          <View style={styles.warningToastTextContainer}>
            <View>
              <BaseText
                style={styles.warningToastTitle}
                variant={BaseTextVariant.lHeadingSM}
              >
                {text1}
              </BaseText>
              <BaseText variant={BaseTextVariant.lBodyXS}>{text2}</BaseText>
            </View>
            <View style={styles.warningToastActionsContainer}>
              <Link
                style={styles.warningToastdismissableAction}
                variant={BaseTextVariant.sBodySM}
                onPress={() => RNToastMessage.hide()}
              >
                {strings('toast.show_warning_toast.remind_me_later')}
              </Link>
              <Link
                variant={BaseTextVariant.sBodySMBold}
                onPress={props.action}
              >
                {props.actionText}
              </Link>
            </View>
          </View>
        </>
      </BaseToast>
    );
  },
});

const Toast: IToast = () => {
  const theme = useAppThemeFromContext();
  const { styles } = useStyles(styleSheet, {});
  return (
    <RNToastMessage
      position="bottom"
      config={toastConfig(styles, theme.colors)}
    />
  );
};

Toast.showWarningToast = ({ title, message, action, actionText }) => {
  RNToastMessage.show({
    type: 'warningToast',
    text1: title,
    text2: message,
    autoHide: false,
    props: {
      action,
      actionText,
    },
  });
};

export default Toast;
