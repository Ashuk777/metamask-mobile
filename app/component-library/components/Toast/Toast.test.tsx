/* eslint-disable no-console */
/* @jest-environment jsdom */
import React from 'react';
import { shallow } from 'enzyme';
import Toast from './';

describe('Toast', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<Toast />);
    expect(wrapper).toMatchSnapshot();
  });
});
