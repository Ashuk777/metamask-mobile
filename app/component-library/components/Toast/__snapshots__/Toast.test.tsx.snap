// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should render correctly 1`] = `
<Toast
  config={
    Object {
      "warningToast": [Function],
    }
  }
  position="bottom"
/>
`;
